src/vos/vos_common.c

```diff
@@ -133,24 +133,21 @@ vos_tx_publish(struct dtx_handle *dth, int err)
 
 again:
 	for (i = 0, rc = 0;
-	     i < dth->dth_modification_cnt && (rc == 0 || err != 0); i++) {
+	     i < dth->dth_rsrvd_cnt && (rc == 0 || err != 0); i++) {
 		dru = &dth->dth_rsrvds[i];
-		if (dru->dru_scm != NULL) {
-			rc = vos_publish_scm(cont, dru->dru_scm, err == 0);
-
-			/* FIXME: Currently, vos_publish_blocks() will release
-			 *	  reserved information in 'dru_nvme_list' from
-			 *	  DRAM. So if vos_publish_blocks() failed, the
-			 *	  former published ones cannot be rollback. It
-			 *	  will be handled later.
-			 */
-			if (rc == 0 || err != 0)
-				rc = vos_publish_blocks(cont, &dru->dru_nvme,
-							err == 0,
-							VOS_IOS_GENERIC);
-			if (err != 0)
-				D_FREE(dru->dru_scm);
-		}
+		rc = vos_publish_scm(cont, dru->dru_scm, err == 0);
+
+		/* FIXME: Currently, vos_publish_blocks() will release
+		 *	  reserved information in 'dru_nvme_list' from
+		 *	  DRAM. So if vos_publish_blocks() failed, the
+		 *	  former published ones cannot be rollback. It
+		 *	  will be handled later.
+		 */
```

Comment:
This comment is actually misleading, if one vos_publish_blocks() failed, nothing will be published because transaction will be aborted.

The real problem could be described as: If 'publish' (for either SCM or NVMe) failed, 'cancel' actually needs be called to cancel the in-memory reservation, but that requires reconstructing (or preserving) the reserved information for the input of cancel call, so we choose to don't call cancel for such case, that'll cause temporal space leaking before in-memory reserve information synced with persistent allocation heap on next restart.

Reply:
I will refresh the comment.

Reply:
This looks better, Thanks. As I mentioned, it's not specific to vos_publish_blocks(), vos_publish_scm() has the same issue.